File: engine/blocks/intake.py
"""Intake — load and validate the user's inputs (the onboarding answers).

The real file (`data/intake.json`) is personal and gitignored. `intake.example.json` shows the
blank shape; `intake.sample.json` is a filled synthetic example used as the cold-start fallback.
Nothing here is invented: every field is something the user gave.
"""

from __future__ import annotations

from pathlib import Path

from pydantic import BaseModel

from ..config import DATA_DIR


class ContentIdea(BaseModel):
    topic: str
    take: str = ""
    scene: str = ""
    number: str = ""  # leave blank if no real number was given
    lesson: str = ""
    only_you: str = ""
    mechanism: str = ""  # how it actually works (real specifics)
    proof: list[str] = []  # receipts to attach / cite
    close: str = ""


class Online(BaseModel):
    linkedin: str = ""
    x: str = ""
    other: str = ""
    existing_posts: str = ""
    cold_start: bool = True


class Docs(BaseModel):
    resume: str = ""  # plain text
    portfolio: str = ""


class Typed(BaseModel):
    identity: str = ""
    known_for: str = ""
    background: str = ""
    beliefs: str = ""
    lessons: str = ""


class Voice(BaseModel):
    answers: dict[str, str] = {}  # question -> raw, unpolished answer
    # real work you're proud of (essays, chats, posts) — the strongest voice corpus
    writing_samples: list[str] = []
    tone_words: list[str] = []
    look: str = ""
    sentence_length: str = ""
    banned: list[str] = []
    signatures: list[str] = []
    emojis: str = ""  # e.g. "none"
    humor: str = ""  # the onboarding A/B pick, e.g. "dry & deadpan"
    shape: str = ""  # the onboarding A/B pick, e.g. "short & punchy"
    notes: str = ""


class Audience(BaseModel):
    writing_for: str = ""
    goal: str = ""
    play_to: str = ""


class OutputPrefs(BaseModel):
    channels: list[str] = ["LinkedIn"]
    length: str = ""
    format: str = ""
    hard_nevers: list[str] = []
    off_limits: str = ""


class Intake(BaseModel):
    name: str
    idea: ContentIdea
    online: Online = Online()
    docs: Docs = Docs()
    typed: Typed = Typed()
    voice: Voice = Voice()
    audience: Audience = Audience()
    output: OutputPrefs = OutputPrefs()


def load_intake(path: Path | None = None) -> Intake:
    if path is None:
        # Personal intake.json wins if present; otherwise fall back to the committed
        # synthetic sample so a fresh checkout runs cold with no setup.
        primary = DATA_DIR / "intake.json"
        path = primary if primary.exists() else (DATA_DIR / "intake.sample.json")
    if not path.exists():
        raise FileNotFoundError(f"intake not found at {path} — see data/intake.example.json")
    return Intake.model_validate_json(path.read_text(encoding="utf-8"))

File: web/lib/intake.ts
// 1:1 mirror of engine/blocks/intake.py (pydantic Intake) and data/intake.example.json.
// Keep this in sync with the Python model — it is the single source of truth the
// onboarding screens assemble and PUT/PATCH to the API.

export interface ContentIdea {
  topic: string; // required by the engine
  take: string;
  scene: string;
  number: string; // "" if no real number
  lesson: string;
  only_you: string;
  mechanism: string;
  proof: string[];
  close: string;
}
export interface Online {
  linkedin: string;
  x: string;
  other: string;
  existing_posts: string;
  cold_start: boolean;
}
export interface Docs {
  resume: string;
  portfolio: string;
}
export interface Typed {
  identity: string;
  known_for: string;
  background: string;
  beliefs: string;
  lessons: string;
}
export interface Voice {
  answers: Record<string, string>;
  tone_words: string[];
  look: string;
  sentence_length: string;
  banned: string[];
  signatures: string[];
  emojis: string;
  notes: string;
}
export interface Audience {
  writing_for: string;
  goal: string;
  play_to: string;
}
export interface OutputPrefs {
  channels: string[];
  length: string;
  format: string;
  hard_nevers: string[];
  off_limits: string;
}
export interface Intake {
  name: string; // required by the engine
  idea: ContentIdea;
  online: Online;
  docs: Docs;
  typed: Typed;
  voice: Voice;
  audience: Audience;
  output: OutputPrefs;
}

export function emptyIntake(): Intake {
  return {
    name: "",
    idea: {
      topic: "",
      take: "",
      scene: "",
      number: "",
      lesson: "",
      only_you: "",
      mechanism: "",
      proof: [],
      close: "",
    },
    online: { linkedin: "", x: "", other: "", existing_posts: "", cold_start: true },
    docs: { resume: "", portfolio: "" },
    typed: { identity: "", known_for: "", background: "", beliefs: "", lessons: "" },
    voice: {
      answers: {},
      tone_words: [],
      look: "",
      sentence_length: "",
      banned: [],
      signatures: [],
      emojis: "",
      notes: "",
    },
    audience: { writing_for: "", goal: "", play_to: "" },
    output: { channels: ["LinkedIn"], length: "", format: "", hard_nevers: [], off_limits: "" },
  };
}

// Truthful onboarding completion (0..1): the share of the fields onboarding asks for
// that are actually filled. Powers the "voice tuned %" meter — no faked numbers.
export function completion(i: Intake): number {
  const filled = (s: string) => s.trim().length > 0;
  const checks: boolean[] = [
    filled(i.name),
    filled(i.online.linkedin) || filled(i.online.x),
    filled(i.docs.resume),
    filled(i.typed.identity),
    filled(i.typed.known_for),
    filled(i.typed.beliefs),
    Object.values(i.voice.answers).some(filled),
    i.voice.tone_words.length > 0,
    filled(i.voice.sentence_length),
  ];
  const done = checks.filter(Boolean).length;
  return done / checks.length;
}
